libavformat/rtcenc.c

```diff
@@ -0,0 +1,2226 @@
+ *
+ * If the profile is less than 0, the function considers the profile as baseline.
+ * It may need to parse the profile from SPS/PPS. This situation occurs when ingesting
+ * desktop and transcoding.
+ *
+ * @param s Pointer to the AVFormatContext
+ * @returns Returns 0 if successful or AVERROR_xxx in case of an error.
+ */
+static int parse_codec(AVFormatContext *s)
+{
+    int i, ret;
+    RTCContext *rtc = s->priv_data;
+
+    for (i = 0; i < s->nb_streams; i++) {
+        AVCodecParameters *par = s->streams[i]->codecpar;
+        const AVCodecDescriptor *desc = avcodec_descriptor_get(par->codec_id);
+        switch (par->codec_type) {
+        case AVMEDIA_TYPE_VIDEO:
+            if (rtc->video_par) {
+                av_log(s, AV_LOG_ERROR, "Only one video stream is supported by RTC\n");
+                return AVERROR(EINVAL);
+            }
+            rtc->video_par = par;
+
+            if (par->codec_id != AV_CODEC_ID_H264) {
+                av_log(s, AV_LOG_ERROR, "Unsupported video codec %s by RTC, choose h264\n",
+                       desc ? desc->name : "unknown");
+                return AVERROR_PATCHWELCOME;
+            }
+
+            if (par->video_delay > 0) {
+                av_log(s, AV_LOG_ERROR, "Unsupported B frames by RTC\n");
+                return AVERROR_PATCHWELCOME;
+            }
+
+            ret = isom_read_avcc(s, par->extradata, par->extradata_size);
+            if (ret < 0) {
+                av_log(s, AV_LOG_ERROR, "Failed to parse SPS/PPS from extradata\n");
+                return ret;
+            }
+            break;
+        case AVMEDIA_TYPE_AUDIO:
+            if (rtc->audio_par) {
+                av_log(s, AV_LOG_ERROR, "Only one audio stream is supported by RTC\n");
+                return AVERROR(EINVAL);
+            }
+            rtc->audio_par = par;
+
+            if (par->codec_id != AV_CODEC_ID_OPUS) {
+                av_log(s, AV_LOG_ERROR, "Unsupported audio codec %s by RTC, choose opus\n",
+                    desc ? desc->name : "unknown");
+                return AVERROR_PATCHWELCOME;
+            }
+
+            if (par->ch_layout.nb_channels != 2) {
+                av_log(s, AV_LOG_ERROR, "Unsupported audio channels %d by RTC, choose stereo\n",
+                    par->ch_layout.nb_channels);
+                return AVERROR_PATCHWELCOME;
+            }
+
+            if (par->sample_rate != 48000) {
+                av_log(s, AV_LOG_ERROR, "Unsupported audio sample rate %d by RTC, choose 48000\n", par->sample_rate);
+                return AVERROR_PATCHWELCOME;
+            }
+            break;
+        default:
+            av_log(s, AV_LOG_ERROR, "Codec type '%s' for stream %d is not supported by RTC\n",
+                   av_get_media_type_string(par->codec_type), i);
+            return AVERROR_PATCHWELCOME;
+        }
+    }
+
+    return 0;
+}
+
+/**
+ * Generate SDP offer according to the codec parameters, DTLS and ICE information.
```

Comment:
Derek Buitenhuis:

```text
> +/**
> + * Generate SDP offer according to the codec parameters, DTLS and ICE information.
> + * The below is an example of SDP offer:
```

This comment and function need some serious work:
  * The example and comment explain exactly nothing about the SDP text being generated. It
    is an opaque blob.
  * It is illegal SDP.
  * We already have SDP facilties you should use.